A particularly nasty bug that I fixed myself! (all the while learning about GitHub)

jeremy-jameson opened the pull request.

Commit:
#6 - Change "Title" attribute to "title" to avoid invalid XHTML

jeremy-jameson committed:
#8 - Sort post categories in ascending order by default

jeremy-jameson committed:
#9 - Comment out CSS rule that refers to obsolete file (/Subtext.Web/Skins/Haackify/Images/bottom.gif)

jeremy-jameson committed:
#10 - Fix bug when a tag is specified using a relative URL instead of an absolute URL

jeremy-jameson committed:
#10 - Allow tags on blog posts to refer to Subtext blog (rather than, say, Technorati) by supporting tag URLs without "/default.aspx" on the end

jeremy-jameson committed:
#11 - Expose the "Entry" entity object in the ViewPost control (so that it is available to skin controls that want to format the fields differently or show additional fields)

jeremy-jameson committed:
#12 - Fix bug where "Submit" button would show "undefined" after partial page update (e.g. when custom CAPTCHA control failed validation)

jeremy-jameson committed:
#14 - Avoid SecurityException when category is not found and Subtext is running in Medium trust

jeremy-jameson committed:
#15 - Avoid error when post is not found and skin control contains code blocks (i.e. "<% ... %>")

jeremy-jameson committed:
#16 - Allow errors in Lucene.Net.QueryParsers.QueryParser to be avoided by disabling the full text search engine in Web.config

jeremy-jameson committed:
#17 - Allow server-relative paths to be specified for CSS files in skin.config file (and treat these CSS references as "remote")

jeremy-jameson committed:
#18 - jQuery references need to specify "https://" (not "http://") when browsing Subtext pages using HTTPS

jeremy-jameson committed:
#19 - Category links do not specify correct URLs when browsing Subtext blog pages using HTTPS

jeremy-jameson committed:
#20 - Fix numerous issues with pingback functionality in Subtext:
- Error in PROD (i.e. "System.Web.HttpException: The file '/blog/jjameson/Services/Pingback.aspx' does not exist.")
- The "pingback" URL specified in the <link> head element should be an absolute URL (not a relative URL) according to the Pingback 1.0 specification (http://www.hixie.ch/specs/pingback/pingback)
- According to the current Subtext routing functionality, the "pingback" URL needs to include the ID of the post (e.g. "http://www.technologytoolbox.com/blog/jjameson/Services/Pingback/315.aspx")

haacked commented on src/Subtext.Framework/Infrastructure/Installation/Scripts/StoredProcedures.sql:
```diff
@@ -1126,6 +1126,8 @@ WHERE (c.CategoryID = @CategoryID OR @CategoryID IS NULL)
     AND (c.BlogId = @BlogId OR @BlogId IS NULL)
     AND c.Active <> CASE @IsActive WHEN 0 THEN -1 else 0 END
+ORDER BY
+    c.Title
```
Let's be explicit about ordering. This should probably be:

```
ORDER BY
    c.Title ASC
```

jeremy-jameson replied:
Done

haacked commented on src/Subtext.Framework/Infrastructure/Routing/SubtextRouteExtensions.cs:
```diff
@@ -233,9 +233,22 @@ public static Uri ToFullyQualifiedUrl(this VirtualPath virtualPath, Blog blog, s
-                builder.Port = HttpContext.Current.Request.Url.Port;
+                if (HttpContext.Current.Request.IsSecureConnection == true)
+                {
+                    builder.Scheme = "https";
```
Nice catch! Do we have unit tests for this method anywhere? I didn't see any updated tests.

jeremy-jameson replied:
No...while I managed to get the unit tests working in the 2.5 solution (i.e. Visual Studio 2008 and MbUnit), I haven't had any success thus far with the current 2.6 solution. MbUnit just barfs on me (complaining about the version of the .NET Framework being used).

Note that I'm a Visual Studio "Team Suite" kind of guy, so I know next to nothing about MbUnit. I'd love to get it working -- and subsequently write tests first when making changes or enhancements -- but I'm going to need a little help to make that happen.

Regardless of my current blocking issue with MbUnit, for this particular fix would it even be possible to write a unit test, given the call to HttpContext.Current? Sorry if this seems ignorant, but in my past experience I've always passed in an HttpContextBase parameter to methods when doing TDD.

haacked replied:
Ok, don't worry about it for now. :) Yeah, this would be hard to test as it stands. But that method would get annoying if we had to pass in an `HttpContextBase` parameter too. Though that's probably the right thing to do. :) I'll assume you tested this thoroughly.

haacked commented on src/Subtext.Framework/Tracking/TrackBackNotificationProxy.cs:
```diff
@@ -72,7 +72,7 @@ public class TrackBackNotificationProxy
                 }
             }
-            return true;
+            return false;
```
Ummm, whoops! Heh. Nice catch.

haacked commented on src/Subtext.Web/aspx/Admin/WebUI/AdminPageTemplate.Master:
```diff
@@ -4,9 +4,9 @@
 <head runat="server">
     <title>Subtext - Admin</title>
-    <script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.6.1/jquery.min.js"></script>
+    <script type="text/javascript" src="<%= Request.IsSecureConnection ? "https" : "http" %>://ajax.googleapis.com/ajax/libs/jquery/1.6.1/jquery.min.js"></script>
```
Actually, we should simply use a [protocol relative URL/network path reference](http://paulirish.com/2010/the-protocol-relative-url/).

For example:

```
src="//ajax.googleapis.com/..."
```

That removes the need do use `<%= %>` blocks here and is cleaner.

jeremy-jameson replied:
Wow...learn something new every day!

I didn't even know that was possible (before reading your comment, I would have assumed a URL beginning with "//" would be interpreted as a server-relative URL).

You are absolutely right, that would be much cleaner. I'll make the changes now.

haacked replied:
Great! Thanks so much for chipping in!

haacked commented on src/Subtext.Web/aspx/HostAdmin/styles/Style.css:
```diff
@@ -560,7 +560,7 @@ img
 #bottom
 {
-	background: url(/Subtext.Web/Skins/Haackify/Images/bottom.gif) no-repeat ;
+	/* background: url(/Subtext.Web/Skins/Haackify/Images/bottom.gif) no-repeat ; */
```
Ack! How'd that get in there! ;)

jeremy-jameson committed:
#8 - Explicitly specify "ascending" sort order (per Phil's request)

jeremy-jameson committed:
#18 - Refine method for specifying "http://" or "https://" for jQuery files (per Phil's feedback)

jeremy-jameson committed:
Merge changes from upstream (https://github.com/haacked/Subtext)

jeremy-jameson committed:
Upgrade log4net to version 1.2.11

- Includes new AdoNetAppender.ConnectionStringName property
- Reference: https://issues.apache.org/jira/browse/LOG4NET-88

jeremy-jameson committed:
Upgrade solution to Visual Studio 2015

- Add NuGet packages for ASP.NET MVC 3 to Framework, Web, and UnitTests projects

jeremy-jameson committed:
Add blog skin and related files for Technology Toolbox website

jeremy-jameson committed:
Tweak messages displayed to blog readers

jeremy-jameson committed:
Expose EntryViewModel (instead of Entry) in ViewPost control (to be consistent with other controls, e.g. PostComment)

jeremy-jameson committed:
Fix merge bug

jeremy-jameson committed:
Remove blog skin and related files for Technology Toolbox website

This reverts commit 3b2702beb1a720b739c64bb462e824bd95ae6c3a.

jeremy-jameson committed:
Add ASP.NET MVC project type GUIDs to SubText.Web.csproj

- Enables the "F5" experience in Visual Studio
- {E53F8FEA-EAE0-44A6-8774-FFD645390401} - ASP.NET MVC 3
- {349C5851-65DF-11DA-9384-00065B846F21} - ASP.NET MVC 5
- Note that ASP.NET MVC 5 is required to enable the "Web" tab in project properties (at least when using Visual Studio 2017)

jeremy-jameson committed:
Add .vs/ and App_Data/ to .gitignore

jeremy-jameson committed:
Upgrade assemblies for SQL Server Management Objects (SMO)

- Fixes #1
- The new versions are from SQL Server 2008 R2 Express Edition (11.0.0.0)

jeremy-jameson committed:
Fix bug in unit tests due to time zone differences

- Fixes #2 (issue where three unit tests failed when running on machines *not* in Pacific time zone)

jeremy-jameson committed:
Avoid ArgumentOutOfRangeException when browsing to ".../blog"

Handle special case where a single blog is hosted at something like "/blog/jjameson" and a user hacks the URL to specify "/blog"

jeremy-jameson committed:
Fix another bug when attempting to redirect from "/blog" to "/blog/jjameson"

jeremy-jameson committed:
Support "placeholder" Gravatar image in blog skins

Only change the Visible property of the Gravatar image to true when an email address is specified in the comment. This allows a skin to specify a default Gravatar image (e.g. PlaceHolderImage="~/Skins/TechnologyToolbox1/Images/Silhouette-1.jpg") but also choose to hide the default image when no email address is available to retrieve a Gravatar (i.e. by specifying Visible="false" on the Image control).

jeremy-jameson committed:
Fix HTTP 404 when browsing to ".../blog/jjameson/tags"

jeremy-jameson committed:
jQuery references need to specify "https://" (not "http://") when browsing Subtext pages using HTTPS

jeremy-jameson committed:
Change web project to use IIS Express

jeremy-jameson committed:
Support alternate URLs without changing the original hostname

- Fix #3
- This is required to support the new **staging** deployment slot for technologytoolbox.com

jeremy-jameson committed:
Upgrade test project to .NET Framework 4.5

This is a prerequisite for installing the latest MSTest.TestFramework NuGet package.

jeremy-jameson committed:
Reinstall NuGet packages

- Update-Package log4net -Reinstall
- Update-Package Ninject -Reinstall
- Update-Package xmlrpcnet -Reinstall

This resolves a warning in Visual Studio after updating test project to .NET 4.5.

jeremy-jameson committed:
Install MSTest NuGet packages

- Install-Package MSTest.TestFramework
- Install-Package MSTest.TestAdapter
- Install-Package Microsoft.NET.Test.Sdk

jeremy-jameson committed:
Convert simplest unit test from MbUnit to MSTest

jeremy-jameson committed:
Convert tests from MbUnit to MSTest (first pass)

jeremy-jameson committed:
Convert database integration tests from MbUnit to MSTest

jeremy-jameson committed:
Convert "multiple culture" tests from MbUnit to MSTest

jeremy-jameson committed:
Convert "TypeFixture" test from MbUnit to MSTest

- While not as "elegant" as the original EmailProvider tests, this is "good enough" for the current test scenarios

jeremy-jameson committed:
Remove MbUnit references from UnitTests.Subtext project

jeremy-jameson committed:
Use SQL Server Express LocalDB for running tests

- This is required to avoid errors when building the solution on Azure DevOps hosted agents

jeremy-jameson committed:
Remove hack added to get test to pass in Visual Studio 2019

jeremy-jameson committed:
Add Microsoft.SqlServer.SqlClrProvider.dll

This is necessary to avoid an error when running the tests in specific environments (e.g. EXT-VS2017-DEV3).

This assembly is installed in the GAC in some environments (e.g. EXT-VS2013-DEV1) -- presumably due to installing the corresponding version of SQL Server Express.

jeremy-jameson committed:
Merge pull request #7 from technology-toolbox/users/jjameson/dev

Convert tests from MbUnit to MSTest

jeremy-jameson committed:
Refine messages for test assertions

- Issue #8
- These two tests intermittently fail when executed on Azure DevOps hosted agents (but consistently succeed in various local environments). Change the messages to be unique to identify which assertions are actually failing.

jeremy-jameson committed:
Verify entries before verifying order of results

- Issue #8

jeremy-jameson committed:
Explicitly set the "date published" property on entries

- Fix #8
- This prevents intermittent test failures depending on the speed at which the code is executed (i.e. two entries having the same value in subtext_Content.DatePublishedUtc).

jeremy-jameson committed:
Merge pull request #9 from technology-toolbox/users/jjameson/dev

Fix intermittent test failures in Azure DevOps build pipeline

jeremy-jameson committed:
Fix category URLs in sitemap

- Fixes #10
- Fixing the bug was easy...trying to create a corresponding unit test turned out to be a nightmare (I eventually punted this due to time constraints)
- The new test currently throws a NullReferenceException in Transformer.GetLinkFromLinkCategory
- TODO: Determine why UrlHelper.GetVirtualPath returns null for category route

jeremy-jameson committed:
Fix archive URLs in sitemap

    - Fixes #11
    - Fixing the bug was easy...but creating a corresponding unit test...not so much (similar to previous bug fix for category URLs)

jeremy-jameson committed:
Merge pull request #12 from technology-toolbox/users/jjameson/dev

Fix URLs in sitemap.ashx

dependabot[bot] committed:
Bump DotNetZip from 1.9.1.8 to 1.11.0 in /src/Subtext.Web

Bumps [DotNetZip](https://github.com/haf/DotNetZip.Semverd) from 1.9.1.8 to 1.11.0.
- [Release notes](https://github.com/haf/DotNetZip.Semverd/releases)
- [Commits](https://github.com/haf/DotNetZip.Semverd/commits)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump DotNetZip from 1.9.1.8 to 1.11.0 in /src/SubtextUpgrader

Bumps [DotNetZip](https://github.com/haf/DotNetZip.Semverd) from 1.9.1.8 to 1.11.0.
- [Release notes](https://github.com/haf/DotNetZip.Semverd/releases)
- [Commits](https://github.com/haf/DotNetZip.Semverd/commits)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump log4net from 1.2.11 to 2.0.10 in /src/Subtext.Web

Bumps [log4net](https://github.com/apache/logging-log4net) from 1.2.11 to 2.0.10.
- [Release notes](https://github.com/apache/logging-log4net/releases)
- [Changelog](https://github.com/apache/logging-log4net/blob/master/ReleaseInstructions.txt)
- [Commits](https://github.com/apache/logging-log4net/compare/rel/1.2.11...rel/2.0.10)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump log4net from 1.2.11 to 2.0.10 in /src/Subtext.Framework

Bumps [log4net](https://github.com/apache/logging-log4net) from 1.2.11 to 2.0.10.
- [Release notes](https://github.com/apache/logging-log4net/releases)
- [Changelog](https://github.com/apache/logging-log4net/blob/master/ReleaseInstructions.txt)
- [Commits](https://github.com/apache/logging-log4net/compare/rel/1.2.11...rel/2.0.10)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump log4net from 1.2.11 to 2.0.10 in /src/UnitTests.Subtext

Bumps [log4net](https://github.com/apache/logging-log4net) from 1.2.11 to 2.0.10.
- [Release notes](https://github.com/apache/logging-log4net/releases)
- [Changelog](https://github.com/apache/logging-log4net/blob/master/ReleaseInstructions.txt)
- [Commits](https://github.com/apache/logging-log4net/compare/rel/1.2.11...rel/2.0.10)

Signed-off-by: dependabot[bot] <support@github.com>
